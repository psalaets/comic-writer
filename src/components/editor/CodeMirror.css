/******************************************************************************/
/* CodeMirror                                                                 */
/******************************************************************************/
/*
 * 1. The superflurous attribute selector is here to out-specify a codemirror
 *    selector.The codemirror selector is `.CodeMirror-wrap pre` (011) and this
 *    is 020
 * 2. the width in "X's" on each side.
 * 3. Each Line of code.
 * 4. Wraps the text in each line's gutter value.
 * 5. Turning off "focus" outline for the editor.
 *    TODO: What's the right way to make this a11y without it looking like total
 *          shit?
 * 6. Resetting z-index to creat a new stacking context for this Component.
*/

/* Block */
/******************************************************************************/
.c-codemirror {
  background-color: var(--color-utility-pure);
  border-left: 1px solid var(--color-utility-milk);
  border-right: 1px solid var(--color-utility-milk);
  height: 100%;
  margin-left: auto;
  margin-right: auto;
  padding-left: 1.5ex; /* 2 */
  padding-right: 3ex; /* 2 */
  width: 100%;
  z-index: 0; /* 6 */
}

/* Misc */
/******************************************************************************/
.CodeMirror-placeholder[class] { /* 1 */
  color: var(--color-utility-dove);
}

.CodeMirror-line[class] {} /* 3 */

.CodeMirror-gutter-elt { /* 4 */
  align-items: flex-start;
  bottom: 0;
  display: flex;
  top: 0;
}

.CodeMirror-gutter-wrapper {
  bottom: 0;
  top: 0;
}

.CodeMirror *:focus { /* 5 */
  outline: none;
}

/* panel count line widget */
.panel-count {
  color: var(--color-utility-dove);
}
