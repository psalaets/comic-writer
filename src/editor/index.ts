import * as actions from './actions';
import reducer from './reducer';
import * as selectors from './selectors';

export { default as Editor } from './components/Editor';

export {
  actions,
  reducer,
  selectors
};
