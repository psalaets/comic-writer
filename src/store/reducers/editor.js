import {
  CHANGE_SOURCE,
  LOAD_SCRIPT_COMPLETED
} from '../action-types';

export default function editorReducer(state, action) {
  state = state || {
    source: ''
  };

  switch (action.type) {
    case LOAD_SCRIPT_COMPLETED: {
      return {
        source: transformMarkdown(action.payload.source)
      };
    }
    case CHANGE_SOURCE: {
      return {
        source: transformMarkdown(action.payload.source)
      };
    }
    default:
      return state;
  }
}

// exported for testing purposes
export function transformMarkdown(value) {
  let pageNumber = 0;
  let panelNumber = 0;

  const lines = value.split(/\n/);
  const newValue = lines
    .map(line => {
      const isPage = line.match(/^page/i);
      const isPanel = line.match(/^panel/i);

      if (isPage) {
        pageNumber += 1;
        panelNumber = 0;
        return `Page ${pageNumber}`;
      } else if (isPanel) {
        panelNumber += 1;
        return `Panel ${panelNumber}`;
      } else {
        return line;
      }
    })
    .join('\n');

  return newValue;
}
