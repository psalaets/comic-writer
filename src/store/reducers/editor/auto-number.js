export default function createAutoNumber() {
  let pagesInComic = 0;
  let panelsInPage = 0;

  return function autoNumber(lineClassification) {
    const { line, type, count } = lineClassification;

    if (type === 'single-page') {
      pagesInComic += 1;
      panelsInPage = 0;

      return `Page ${pagesInComic}`;
    }

    if (type ===  'multi-page') {
      const startPage = pagesInComic + 1;
      panelsInPage = 0;

      if (count > 1) {
        pagesInComic += count;

        return `Pages ${startPage}-${pagesInComic}`;
      } else {
        return line;
      }
    }

    if (type === 'invalid-page-range') {
      pagesInComic += 2;
      panelsInPage = 0;

      return line;
    }

    if (type === 'partial-page-range') {
      pagesInComic += 1;
      panelsInPage = 0;

      return `Pages ${pagesInComic}-`;
    } else if (type === 'panel') {
      panelsInPage += 1;

      return `Panel ${panelsInPage}`;
    } else {
      return line;
    }
  };
}
