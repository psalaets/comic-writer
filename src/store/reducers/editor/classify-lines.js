const PAGE_EXPANSION_PATTERN = /^page *$/i;
const PAGES_EXPANSION_PATTERN = /^pages *$/i;
const SPREAD_EXPANSION_PATTERN = /^spread$/i;

const PANEL_PATTERN = /^panel/i;
const SINGLE_PAGE_PATTERN = /^pages? +\d{1,}$/i;
const PAGE_RANGE_PATTERN = /^pages? +(\d{1,})-(\d{1,})$/i;
const PARTIAL_PAGE_RANGE_PATTERN = /^pages? \d{1,}-$/i;

export default function classifyLines(cursorLine) {
  return function classify(line, lineNumber) {
    const cursorOnThisLine = lineNumber === cursorLine;

    if (line.match(PAGE_EXPANSION_PATTERN)) {
      return cursorOnThisLine ? regularLine(line) : singlePageLine(line);
    }

    if (line.match(PAGES_EXPANSION_PATTERN)) {
      return cursorOnThisLine ? regularLine(line) : multiPageLine(line, 2);
    }

    if (line.match(SINGLE_PAGE_PATTERN)) {
      return singlePageLine(line);
    }

    const pageRange = line.match(PAGE_RANGE_PATTERN);
    if (pageRange) {
      let start = parseInt(pageRange[1], 10);
      let end = parseInt(pageRange[2], 10);

      if (start < end) {
        return multiPageLine(line, 1 + end - start);
      }

      // handle weird ranges
      if (start > end) {
        return cursorOnThisLine
          ? invalidPageRangeLine(line)
          // with cursor gone it's like a 2 pager
          : multiPageLine(line, 2);
      }

      // start === end
      return singlePageLine(line);
    }

    if (line.match(PARTIAL_PAGE_RANGE_PATTERN)) {
      return cursorOnThisLine
        ? partialPageRangeLine(line)
        : singlePageLine(line);
    }

    if (line.match(PANEL_PATTERN)) {
      return panelLine(line);
    }

    if (line.match(SPREAD_EXPANSION_PATTERN)) {
      return cursorOnThisLine ? regularLine(line) : multiPageLine(line, 2);
    }

    return regularLine(line);
  };
}

function regularLine(line) {
  return {
    type: 'regular',
    line
  };
}

function singlePageLine(line) {
  return {
    type: 'single-page',
    count: 1,
    line
  };
}

function multiPageLine(line, count) {
  return {
    type: 'multi-page',
    count,
    line
  };
}

// startPage and a dash but no endPage
function partialPageRangeLine(line) {
  return {
    type: 'partial-page-range',
    line
  };
}

// startPage >= endPage
function invalidPageRangeLine(line) {
  return {
    type: 'invalid-page-range',
    line
  };
}

function panelLine(line) {
  return {
    type: 'panel',
    line
  };
}
