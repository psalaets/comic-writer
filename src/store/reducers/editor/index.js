import sourceReducer from './source';
import {
  CHANGE_SOURCE
} from '../../action-types';

export default function editorReducer(state = {}, action) {
  return {
    source: sourceReducer(state.source, action),
    cursor: cursorReducer(state.cursor, action)
  };
}

function cursorReducer(state = 0, action) {
  switch (action.type) {
    case CHANGE_SOURCE:
      return action.payload.cursor;
    default:
      return state;
  }
}