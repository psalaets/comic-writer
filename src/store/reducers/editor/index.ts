import {
  CHANGE_SOURCE,
  LOAD_SCRIPT_COMPLETED,
  EditorActionTypes,
  EditorState
} from '../../types';

const initialState: EditorState = {
  source: ''
};

export default function editorReducer(
  state = initialState,
  action: EditorActionTypes
): EditorState {
  switch (action.type) {
    case LOAD_SCRIPT_COMPLETED: {
      return {
        source: action.payload.source
      };
    }
    case CHANGE_SOURCE: {
      return {
        source: action.payload.source
      };
    }
    default:
      return state;
  }
}
