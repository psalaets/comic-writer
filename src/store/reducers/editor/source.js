import {
  CHANGE_SOURCE
} from '../../action-types';

export default function sourceReducer(state = '', action) {
  switch (action.type) {
    case CHANGE_SOURCE:
      return action.payload.source;
    default:
      return state;
  }
}