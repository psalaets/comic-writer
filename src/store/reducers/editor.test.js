import { transformMarkdown } from './editor';

describe('editor reducer', () => {
  describe('transformMarkdown', () => {
    describe('pages', () => {
      it('first page', () => {
        const input = `## `;

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('adding page at start', () => {
        const input = '## \n## Page 1\npage 1 info';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('adding page at end', () => {
        const input = '## Page 1\npage 1 info\n## ';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('adding page in the middle', () => {
        const input = '## Page 1\npage 1 info\n## \n## Page 2\npage 2 info';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('inserting page at start', () => {
        const input = '## Page 3\npage 3 info\n## Page 1\npage 1 info\n## Page 2\npage 2 info';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('inserting page at end', () => {
        const input = '## Page 2\npage 2 info\n## Page 1\npage 1 info';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('inserting page in middle', () => {
        const input = '## Page 1\npage 1 info\n## Page 3\npage 3 info\n## Page 2\npage 2 info';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('remove page from start', () => {
        const input = '## Page 2\npage 2 info\n## Page 3\npage 3 info';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('remove page from middle', () => {
        const input = '## Page 1\npage 1 info\n## Page 3\npage 3 info';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('inserting content with page count in double digits', () => {
        const tenPages = Array.from(new Array(10), (p, i) => `## Page ${i + 1}`).join('\n');
        const input = tenPages + '\n## \n## \n## ';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });
    });

    describe('panels', () => {
      it('adding panel to page', () => {
        const input = '## Page 1\npage 1 info\n### ';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('adding second panel to page', () => {
        const input = '## Page 1\npage 1 info\n### Panel 1\n### ';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('adding panel at start', () => {
        const input = '## Page 1\npage 1 info\n### \n### Panel 1\npanel 1 info';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('adding panel in middle', () => {
        const input = '## Page 1\n### Panel 1\npanel 1 info\n### \n### Panel 2\npanel 2 info';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('adding panel to second page', () => {
        const input = '## Page 1\n### Panel 1\npanel 1 info\n## Page 2\n### ';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('adding another panel to second page', () => {
        const input = '## Page 1\n### Panel 1\npanel 1 info\n## Page 2\n### Panel 1\n### ';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('pasting panel at start', () => {
        const input = '## Page 1\npage 1 info\n### Panel 2\n### Panel 1\npanel 1 info';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('pasting panel in middle', () => {
        const input = '## Page 1\n### Panel 1\npanel 1 info\n### Panel 3\n### Panel 2\npanel 2 info';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('removing panel from start', () => {
        const input = '## Page 1\n### Panel 2\npanel 2 info';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });

      it('removing panel from middle', () => {
        const input = '## Page 1\n### Panel 1\npanel 1 info\n### Panel 3\npanel 3 info';

        const result = transformMarkdown(input, 0);

        expect(result.value).toMatchSnapshot();
      });
    });
  });
});