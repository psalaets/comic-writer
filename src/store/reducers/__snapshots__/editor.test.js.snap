// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`editor reducer transformMarkdown pages adding page at end 1`] = `
"Page 1
this is page 1 info
Page 2"
`;

exports[`editor reducer transformMarkdown pages adding page at start 1`] = `
"Page 1
Page 2
this is page 1 info"
`;

exports[`editor reducer transformMarkdown pages adding page in the middle 1`] = `
"Page 1
this is page 1 info
Page 2
Page 3
this is page 2 info"
`;

exports[`editor reducer transformMarkdown pages first page 1`] = `"Page 1"`;

exports[`editor reducer transformMarkdown pages inserting content with page count in double digits 1`] = `
"Page 1
Page 2
Page 3
Page 4
Page 5
Page 6
Page 7
Page 8
Page 9
Page 10
Page 11
Page 12
Page 13"
`;

exports[`editor reducer transformMarkdown pages inserting page at end 1`] = `
"Page 1
this is page 2 info
Page 2
this is page 1 info"
`;

exports[`editor reducer transformMarkdown pages inserting page at start 1`] = `
"Page 1
this is page 3 info
Page 2
this is page 1 info
Page 3
this is page 2 info"
`;

exports[`editor reducer transformMarkdown pages inserting page in middle 1`] = `
"Page 1
this is page 1 info
Page 2
this is page 3 info
Page 3
this is page 2 info"
`;

exports[`editor reducer transformMarkdown pages remove page from middle 1`] = `
"Page 1
this is page 1 info
Page 2
this is page 3 info"
`;

exports[`editor reducer transformMarkdown pages remove page from start 1`] = `
"Page 1
this is page 2 info
Page 2
this is page 3 info"
`;

exports[`editor reducer transformMarkdown panels adding another panel to second page 1`] = `
"Page 1
Panel 1
this is panel 1 info
Page 2
Panel 1
Panel 2"
`;

exports[`editor reducer transformMarkdown panels adding panel at start 1`] = `
"Page 1
this is page 1 info
Panel 1
Panel 2
this is panel 1 info"
`;

exports[`editor reducer transformMarkdown panels adding panel in middle 1`] = `
"Page 1
Panel 1
this is panel 1 info
Panel 2
Panel 3
this is panel 2 info"
`;

exports[`editor reducer transformMarkdown panels adding panel to page 1`] = `
"Page 1
this is page 1 info
Panel 1"
`;

exports[`editor reducer transformMarkdown panels adding panel to second page 1`] = `
"Page 1
Panel 1
this is panel 1 info
Page 2
Panel 1"
`;

exports[`editor reducer transformMarkdown panels adding second panel to page 1`] = `
"Page 1
this is page 1 info
Panel 1
Panel 2"
`;

exports[`editor reducer transformMarkdown panels inserting panel at start 1`] = `
"Page 1
this is page 1 info
Panel 1
Panel 2
this is panel 1 info"
`;

exports[`editor reducer transformMarkdown panels inserting panel in middle 1`] = `
"Page 1
Panel 1
this is panel 1 info
Panel 2
Panel 3
this is panel 2 info"
`;

exports[`editor reducer transformMarkdown panels removing panel from middle 1`] = `
"Page 1
Panel 1
this is panel 1 info
Panel 2
this is panel 3 info"
`;

exports[`editor reducer transformMarkdown panels removing panel from start 1`] = `
"Page 1
Panel 1
this is panel 2 info"
`;
